Improve user input validation for URL fields

afercia opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Improved user input validation for URLs

This PR only improves the existing `WPSEO_Option->validate_url()` method. A proper user input validation should include better error reporting, clear identification of the input with the error, etc. These parts will be addressed in future PRs.

## Relevant technical choices:
`WPSEO_Option->validate_url()` is supposed to render a WP error notice when an invalid URL is submitted. However, its logic fails (see https://github.com/Yoast/bugreports/issues/457) and a "success" notice is always displayed, regardless of the URL is valid or invalid:

<img width="646" alt="screenshot 2019-03-06 at 11 17 13" src="https://user-images.githubusercontent.com/1682452/53874710-49474d80-4003-11e9-97bc-65e90fcea188.png">

Seems to me a better approach for _user input validation_ is to:
- first: validate the user input checking if the entered value matches the required data type, in this case: a valid URL
- then sanitize it

For this purpose, I'm using `filter_var()` and `FILTER_VALIDATE_URL`

## Test instructions

`WPSEO_Option->validate_url()` is currently used at these locations:

**SEO > Social:**
case 'facebook_site':
case 'instagram_url':
case 'linkedin_url':
case 'myspace_url':
case 'pinterest_url':
case 'og_default_image':
case 'og_frontpage_image':
case 'youtube_url':
case 'wikipedia_url':

**SEO > Search Appearance > General:**
case 'company_logo':
case 'person_logo':

Test all these cases, first on trunk:
- enter an invalid value e.g. `q w e r t y` with spaces
- see it gets:
  - prepended with `http://`
  - sanitized: the spaces become `%20`
  - as such, it's considered a "valid" URL so there's no error notice
  - instead, a "success" notice appears

Switch to this branch
- repeat the steps above
- see that `q w e r t y` gets first validated to check if it's a valid URL
- an error notice appears
- the URL field stays empty
- then enter a valid URL
- see it gets saved correctly
- then enter again an invalid URL
- an error notice appears
- the input field is filled with the previous valid value

Note: for the image fields:

case 'og_default_image':
case 'og_frontpage_image':
case 'company_logo':
case 'person_logo':

the UI actually doesn't allow to manually enter a value. These fields are readonly. The UI only allows to select an image from the WP Media Library so the URL is always valid (hopefully) and validation is a bit redundant, though it doesn't harm.

Note to test `og_frontpage_image` you have to have your homepage set to your latest posts. See: https://kb.yoast.com/kb/custom-open-graph-tags/

Adjusted 2 related tests that used an _absolute URL path_ e.g. `/test.png` for the og_default_image and the frontpage_image. These aren't valid URLs, and the UI doesn't allow to use them anyways.

* Because we've had some issues with the multisite tests, please also test on multisite.

## UI changes
* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation
* [ ] I have written documentation for this change.

Fixes https://github.com/Yoast/bugreports/issues/457

afercia commented:
Note: `add_settings_error()` exists since WordPress 3.0. We should remove all the `if ( function_exists( 'add_settings_error' ) )` conditionals. For now, I removed only one of them in the part of code touched by this PR.

IreneStr commented on inc/options/class-wpseo-option.php:
```diff
@@ -320,36 +320,44 @@ public function validate_verification_string( $key, $dirty, $old, &$clean ) {
-			}
-			else {
+
+			$submitted_url = trim( htmlspecialchars( $dirty[ $key ] ) );
```
Is this the correct function to use for an URL (because it's not HTML)?

IreneStr commented on inc/options/class-wpseo-option.php:
```diff
@@ -320,36 +320,44 @@ public function validate_verification_string( $key, $dirty, $old, &$clean ) {
-			}
-			else {
+
+			$submitted_url = trim( htmlspecialchars( $dirty[ $key ] ) );
```
Is this the correct function to use for an URL (because it's not HTML)?

afercia replied:
At this point of the code we just need to trim and do some basic encoding, so that `FILTER_VALIDATE_URL` can do is job. The purpose here is to check if the entered URL has a valid format. A full sanitization happens later.

afercia commented:
Added tests.

Saving a new URL or restoring the previous value if the new one is invalid is tested using the existing `test_validate_option()`. Added related data to the data provider.

Added a new method to test submitting an invalid URL correctly adds an error to the WordPress global `$wp_settings_errors`, which is responsible to print out admin notices.

Run `composer install && composer integration-test` and check tests pass.

IreneStr commented on integration-tests/inc/options/test-class-wpseo-option-social.php:
```diff
@@ -72,6 +73,49 @@ public function validate_option_provider() {
+	 *
+	 * @covers WPSEO_Option_Social::validate_url
+	 */
+	public function test_validate_url_adds_settings_error() {
```
I think a better way to test this is to check if add_settings_error is called once (with the arguments you expect). We don't want to use WordPress globals in tests. When we're moving it to a brain monkey test, that wouldn't even be possible.

igorschoester commented:
Acceptance WIP

* In the knowledge graph settings when:
    * You have it saved to a valid organization.
    * You change to a person and change the URL to something invalid, then switch back to organization and hit save.
    * You still get the invalid url for the person. But that is not of interest to the user.

TODO: test in multisite

afercia commented on tests/inc/options/option-social-test.php:
```diff
@@ -0,0 +1,160 @@
+		$clean = array( 'instagram_url' => '' );
+		$dirty = array( 'instagram_url' => 'invalidurl' );
+
+		$instance->validate_url( 'instagram_url', $dirty, '', $clean );
```
I'm not sure I understand what this Brain-Monkey refactored test actually tests. 
The original version was checking the `global $wp_settings_errors` for a specific setting name and error code name, which was a way to check the presence of the specific added error in the global.

To my understanding, this new version expects `add_settings_error()` to be called once with the passed arguments. It doesn't check the actual presence of the added error in the $wp_settings_errors` array.

Aside: I'm not sure `validate_url()` should be called directly, as it's used internally by `validate_option()`.  Also, as it is now, this test seems redundant to me as it basically replicates what `test_validate_option_with_invalid_data()` already tests for (see `validate_option_invalid_data_provider`).
